File: ga_util.py
# Provides simple functions to be used with the genetic algorithm
import numpy as np
from random_util import RandomUtil
from functools import cmp_to_key

class GaUtil():
    rand_util_instance = RandomUtil(0)
    coefficients_count = 0
    def __init__(self, random_util, coefficients_count):
        self.rand_util_instance = random_util
        self.coefficients_count = coefficients_count
    # Combines coefficients from organism1 and organism2
    # Mutates the child and returns the coefficients
    def combine_organisms(self,organism1,organism2,mutation_factor_range):
        coefs1 = np.array(organism1.coefficients)
        coefs2 = np.array(organism2.coefficients)
        child_coefs = (coefs1+coefs2)/2.0 # Average the coefficients

        mutation_vector = self.rand_util_instance.get_random_array(len(child_coefs),
                                                mutation_factor_range[0],mutation_factor_range[1])
        mutation_vector = np.multiply(child_coefs,mutation_vector) # Multiply element-wise    
        child_coefs += mutation_vector # Mutate

        return list(child_coefs)
    
    
    def compare_organisms(self,organism1,organism2):
        if organism1.time_alive < organism2.time_alive:
            return 1
        elif organism1.time_alive > organism2.time_alive:
            return -1
        else:
            return 0
    
    # Returns coefficients gotten from combining random pairs of organisms
    # count - number of coefficients to return
    def get_coeffs_from_best(self, organisms, total_count, no_best, no_random, mutation_factor_range):
        ordered_organisms = sorted(organisms,key=cmp_to_key(self.compare_organisms))
        best_organisms = ordered_organisms[:no_best]

        coeffs = []
        for _ in range(total_count - no_random):
            positions = self.rand_util_instance.get_random_array_discrete(2,0,len(best_organisms))
            coeffs.append(self.combine_organisms(best_organisms[positions[0]],
                                            best_organisms[positions[1]],mutation_factor_range))

        coeffs += self.rand_util_instance.get_random_matrix(
            no_random, self.coefficients_count, [-1000,1000])

        return coeffs

File: genetic_algorithm_cpp/GaUtil.cpp
#include "GaUtil.h"
#include <algorithm>

GaUtil::GaUtil(RandomUtil* rand_util, int coefficients_count)
{
    this->rand_util = rand_util;
    this->coefficients_count = coefficients_count;
}

vector<double> GaUtil::combine_organisms(Organism* organism1, Organism* organism2, double mutation_stddev)
{
    auto coefs1 = organism1->coefficients;
    auto coefs2 = organism2->coefficients;
    
    vector<double> child_coefs = vector<double>(coefs1.size());
    for (size_t i = 0; i < coefs1.size(); i++)
        child_coefs[i] = (coefs1[i] + coefs2[i]) / 2.0;
   
    vector<double> mutation_vector(coefs1.size());
    mutation_vector = rand_util->get_random_array(coefs1.size(),
            0, mutation_stddev, "normal");
    
    for (size_t i = 0; i < mutation_vector.size(); i++)
        child_coefs[i] += mutation_vector[i] * child_coefs[i];
    
    return child_coefs;
}

vector<vector<double>> GaUtil::get_coeffs_from_best(vector<Organism>* organisms, int total_count, double mutation_stddev, bool as_parents)
{
    vector<vector<double>> best_coefs; best_coefs.reserve(total_count);
    vector<int> positions(organisms->size());
    vector<double> scores(organisms->size());
    for (size_t i = 0; i < organisms->size(); i++)
    {
        positions[i] = i;
        scores[i] = max(0.0,organisms->at(i).time_alive-20);
    }

    if (as_parents)
    {
        auto parent_pairs = rand_util->random_choices(positions, scores, 2 * total_count);
        for (size_t i = 0; i < 2 * total_count; i += 2)
        {
            int pos1 = parent_pairs[i];
            int pos2 = parent_pairs[i + 1];
            best_coefs.push_back(combine_organisms(&organisms->at(pos1), &organisms->at(pos2), mutation_stddev));
        }
    }
    else
    {
        auto chosen_organisms = rand_util->random_choices(positions, scores, total_count);
        for (size_t i = 0; i < total_count; i++)
        {
            int pos = chosen_organisms[i];
            best_coefs.push_back(organisms->at(pos).coefficients);
        }
    }
    
    return best_coefs;
}

bool GaUtil::compare_organisms(Organism organism1, Organism organism2)
{
    return organism1.time_alive > organism2.time_alive;
}
